tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -4769,61 +4775,84 @@ Status ConvertPad(OpConverterParams* params) {
   }
   auto pad_data = static_cast<int*>(pads.GetValues());
 
-  std::vector<int32_t> pad_index;
+  std::vector<int32_t> tf_pad_index;
   for (int i = 0; i < nb_dims; i++) {
     if (pad_data[2 * i] != 0 || pad_data[2 * i + 1] != 0) {
-      pad_index.push_back(i);
+      tf_pad_index.push_back(i);
     }
   }
 
   // No padding at all, we should exit
-  if (pad_index.empty()) {
+  if (tf_pad_index.empty()) {
     params->outputs->push_back(inputs.at(0));
     return Status::OK();
   }
 
-  // Only supports padding on less than 2 axis GIE-2579
-  if (pad_index.size() > 2) {
+  // Only supports padding on less equal than 2 axis GIE-2579
+  // TODO(tfeher) This could be improved by adding multiply TRT pad layers
+  if (tf_pad_index.size() > 2) {
     return errors::InvalidArgument(
         "Padding layer does not support padding on > 2");
   }
 
   // Padding on batch dimension is not supported
-  if (pad_index[0] == 0) {
+  if (params->use_implicit_batch && tf_pad_index[0] == 0) {
     return errors::InvalidArgument(
         "Padding layer does not support padding on batch dimension");
   }
 
-  // Not doing the legit thing here. ignoring padding on dim 1 and 3;
-  // TODO(jie): implement pad as uff parser
-  if (pad_index.size() == 2 && pad_index[0] == 0 && pad_index[1] == 3) {
-    return errors::Unimplemented(
-        "Padding layer does not support padding on dimension 1 and 3 yet");
-  }
   if (params->validation_only) return Status::OK();
 
-  bool legit_pad = true;
+  // TRT can only do the padding at the last two dimensions. We transpose the
+  // input tensor if needed.
+  bool transposed_pad = false;
+  std::vector<int> transpose_idx(nb_dims);
+  std::iota(transpose_idx.begin(), transpose_idx.end(), 0);
+
+  // trt_pad_index denotes the actual idx where the padding is performed by TRT.
+  std::vector<int> trt_pad_index{nb_dims - 2, nb_dims - 1};
+  bool trt_pad_index_transposed = false;
+
+  // How many zeros are padded at the last two dimensions.
   nvinfer1::DimsHW pre_padding(0, 0);
   nvinfer1::DimsHW post_padding(0, 0);
 
-  std::vector<int32_t> permuted_pad_index(pad_index);
-  if (pad_index[0] == 1) {
-    legit_pad = false;
-    TF_RETURN_IF_ERROR(params->converter->TransposeTensor(
-        tensor, {0, 3, 2, 1}, &tensor, node_def, "to_pad"));
-    permuted_pad_index[0] = 3;
-  }
-
-  for (size_t i = 0; i < pad_index.size(); i++) {
-    int index = pad_index[i];
-    if (permuted_pad_index[i] == 2) {
-      pre_padding.h() = pad_data[index * 2];
-      post_padding.h() = pad_data[index * 2 + 1];
-    } else if (permuted_pad_index[i] == 3) {
-      pre_padding.w() = pad_data[index * 2];
-      post_padding.w() = pad_data[index * 2 + 1];
+  // Two special cases where we can avoid permutations.
+  if (tf_pad_index.size() == 1 && tf_pad_index[0] == nb_dims - 1) {
+    // In case there is only one dim to pad, and tf_index is the last dim,
+    // then we set trt_idx to be the same dim.
```

Comment:
Indeed, the comment can be improved. I did some refactorization and changed the comments too, please have a look. 

(1) is handled by the initialization of `trt_pad_index`.
(2) No, in this case we only need to set one value. TRT expect the padding values for the last two dimensions. We initialize `pre_padding` (and also `post_padding`) to zeros for both values. We need to change only a single value. There was actually a problem here, because we have to make sure that the single value goes to the right position in the `pre_padding.d `and `post_padding.d` arrays. I have fixed this by introducing the `trt_pre_post_padding_index` array (and removed the` trt_pad_index_transposed` bool variable, which was trying to do something similar, but failed). I have added unit tests to make sure that we handle cases with 1 padding dim correctly.

The code is possibly still more complex than ideal, let me know if you have some ideas to simplify it.